tracer/test/Datadog.Trace.Tests/Propagators/W3CTraceContextPropagatorTests.cs

```diff
@@ -23,204 +25,284 @@ static W3CTraceContextPropagatorTests()
-            ulong spanId = 987654321;
-            var samplingPriority = SamplingPriorityValues.UserKeep;
-            var context = new SpanContext(traceId, spanId, samplingPriority, serviceName: null, null);
+            var context = new SpanContext(
+                traceId: 123456789,
+                spanId: 987654321,
+                SamplingPriorityValues.UserKeep,
+                serviceName: null,
+                "origin");
 
-            // using IHeadersCollection for convenience, but carrier could be any type
             var headers = new Mock<IHeadersCollection>();
 
             W3CPropagator.Inject(context, headers.Object, (carrier, name, value) => carrier.Set(name, value));
 
             headers.Verify(h => h.Set("traceparent", "00-000000000000000000000000075bcd15-000000003ade68b1-01"), Times.Once());
+            headers.Verify(h => h.Set("tracestate", "dd=s:2;o:origin"), Times.Once());
             headers.VerifyNoOtherCalls();
+        }
+
+        [Theory]
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68b1-00", 123456789, 987654321, false, "000000000000000000000000075bcd15", "000000003ade68b1")]
+        [InlineData("00-0000000000000000000000003ade68b1-00000000075bcd15-01", 987654321, 123456789, true, "0000000000000000000000003ade68b1", "00000000075bcd15")]
+        public void TryParseTraceParent(string header, ulong traceId, ulong spanId, bool sampled, string rawTraceId, string rawParentId)
+        {
+            var expected = new W3CTraceParent(
+                traceId,
+                spanId,
+                sampled,
+                rawTraceId,
+                rawParentId);
+
+            W3CTraceContextPropagator.TryParseTraceParent(header, out var traceParent).Should().BeTrue();
+
+            traceParent.Should().BeEquivalentTo(expected);
+        }
+
+        [Theory]
+        [InlineData(null)]                                                      // null
+        [InlineData("")]                                                        // empty
+        [InlineData(" ")]                                                       // whitespace
+        [InlineData("01-000000000000000000000000075bcd15-000000003ade68b1-0")]  // too short
+        [InlineData("010-00000000000000000000000075bcd15-000000003ade68b1-00")] // wrong hyphen location #1
+        [InlineData("01-000000000000000000000000075bcd150-00000003ade68b1-00")] // wrong hyphen location #2
+        [InlineData("01-000000000000000000000000075bcd15-000000003ade68b-100")] // wrong hyphen location #3
+        [InlineData("01-000000000000000000000000075bcd15-000000003ade68b1-00")] // bad version value
+        [InlineData("00-000000000000000000000000075bcd1z-000000003ade68b1-00")] // bad trace id value
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68bx-00")] // bad parent id value
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68b1-02")] // bad sampled value
```

Comment:
> bad version value

I can't see what's wrong with this one, which is on me I'm sure 😄  

> bad sampled value

Feasibly, couldn't we receive sampling values we don't understand in the future? I assume this case is already handled by the RFC, just clarifying

Reply:
Notes on correctness:
Version: The only invalid version string is "FF" or any string that is NOT two hex digits. If it's "00" then it's good and if it's newer then we still proceed, even though we're operating on the "00" spec.

Sampled value: It having the value "02" isn't necessarily _bad_, it just means that the sampled flag is set to to 0, since 02 technically means a bit field of 0b0000_0010 and the sampled flag is the last bit.

Like I noted before, I'm not sure how much of this PR you want to spend correcting our traceparent parsing, but these would be two fixes to our existing implementation 😄

Reply:
Oops, thanks Zach!

Reply:
> Version: ... if it's newer then we still proceed, even though we're operating on the "00" spec.

Really? Even though the only version we understand is `00`? Any future version will potentially have a different syntax.

Reply:
Discussed elsewhere, but the spec does include a section about versioning and trying to parse future versions: https://www.w3.org/TR/trace-context/#versioning-of-traceparent. Thanks, Zach!

Reply:
>> bad version value
> I can't see what's wrong with this one, which is on me I'm sure

The only valid version today is `00`, the first two chars. But I updated my implementation to follow Zach's comments above and allow any version (except `ff` which is explicitly not valid).

>> bad sampled value
> Feasibly, couldn't we receive sampling values we don't understand in the future? I assume this case is already handled by the RFC, just clarifying

Not really. `sampled` is a Boolean flag. But Zach reminded me that the last two chars form an entire byte which can contain more bytes, so I am updating this code to check specifically for the `sampled` flag and ignore any others.

Reply:
Done!